store/s3storeKMS.go

```diff
@@ -0,0 +1,478 @@
+			default:
+				return secretObject{}, false, err
+			}
+		}
+		// generic errors
+		return secretObject{}, false, err
+	}
+
+	raw, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return secretObject{}, false, err
+	}
+
+	var obj secretObject
+	if err := json.Unmarshal(raw, &obj); err != nil {
+		return secretObject{}, false, err
+	}
+
+	return obj, true, nil
+
+}
+
+func (s *S3KMSStore) readObjectById(id SecretId) (secretObject, bool, error) {
+	path := getObjectPath(id)
+	return s.readObject(path)
+}
+
+func (s *S3KMSStore) puts3raw(path string, contents []byte) error {
+	putObjectInput := &s3.PutObjectInput{
+		Bucket:               aws.String(s.bucket),
+		ServerSideEncryption: aws.String(s3.ServerSideEncryptionAwsKms),
+		SSEKMSKeyId:          aws.String(s.KMSKey()),
+		Key:                  aws.String(path),
+		Body:                 bytes.NewReader(contents),
+	}
+
+	_, err := s.svc.PutObject(putObjectInput)
+	return err
+}
+
+func (s *S3KMSStore) readLatestFile(path string) (LatestIndexFile, error) {
+	getObjectInput := &s3.GetObjectInput{
+		Bucket: aws.String(s.bucket),
+		Key:    aws.String(path),
+	}
+
+	resp, err := s.svc.GetObject(getObjectInput)
+
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			if aerr.Code() == "AccessDenied" {
+				// If we're not able to read the latest index for a KMS Key then proceed like it doesn't exist.
+				// We do this because in a chamber secret folder there might be other secrets written with a KMS Key that you don't have access to.
+				return LatestIndexFile{Latest: map[string]LatestValue{}}, nil
+			}
+
+			if aerr.Code() == s3.ErrCodeNoSuchKey {
+				// Index doesn't exist yet, return an empty index
+				return LatestIndexFile{Latest: map[string]LatestValue{}}, nil
+			}
+		}
+		return LatestIndexFile{}, err
+	}
+
+	raw, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return LatestIndexFile{}, err
+	}
+
+	var index LatestIndexFile
+	if err := json.Unmarshal(raw, &index); err != nil {
+		return LatestIndexFile{}, err
+	}
+
+	return index, nil
+}
+
+func (s *S3KMSStore) readLatest(service string) (LatestIndexFile, error) {
+	// Create an empty latest, this will be used to merge together the various KMS Latest Files
+	latestResult := LatestIndexFile{Latest: map[string]LatestValue{}}
+
+	// List all the files that are prefixed with kms and use them as latest.json files for that KMS Key.
+	params := &s3.ListObjectsInput{
+		Bucket:     aws.String(s.bucket),
+		Prefix:     aws.String(fmt.Sprintf("%s/kms", service)),
+	}
+
+	err := s.svc.ListObjectsPages(params, func(page *s3.ListObjectsOutput, lastPage bool) bool {
+		for index := range page.Contents {
+			key_name := *page.Contents[index].Key
+			result, err := s.readLatestFile(key_name)
+			if err != nil {
+				// Panic if we encounter an error reading the index (keeping in mind it handles AccessDenied).
+				panic(fmt.Sprintf("Error reading latest index for KMS Key (%s): %s", key_name, err))
```

Comment:
:(

Reply:
:)

Reply:
To summarise convo, you could:

- Outside call to `ListObjectsPages` do something like `paginationErr := nil`
- Instead of panicking, set that and pretend you hit the last page so you can get out of the inner thing
- Check after the `ListObjectsPages` call for that `paginationErr` and bubble it up if non-nil